qualtran/_infra/adjoint_test.py

```diff
@@ -0,0 +1,147 @@
+    assert (
+        TestSerialCombo().decompose_bloq().debug_text()
+        == """\
+TestAtom('atom0')<0>
+  LeftDangle.reg -> q
+  q -> TestAtom('atom1')<1>.q
+--------------------
+TestAtom('atom1')<1>
+  TestAtom('atom0')<0>.q -> q
+  q -> TestAtom('atom2')<2>.q
+--------------------
+TestAtom('atom2')<2>
+  TestAtom('atom1')<1>.q -> q
+  q -> RightDangle.reg"""
+    )
+
+    # The adjoint reverses the order to atom{2,1,0} and wraps each in `Adjoint`.
+    assert (
+        TestSerialCombo().adjoint().decompose_bloq().debug_text()
+        == """\
+Adjoint(subbloq=TestAtom('atom2'))<0>
+  LeftDangle.reg -> q
+  q -> Adjoint(subbloq=TestAtom('atom1'))<1>.q
+--------------------
+Adjoint(subbloq=TestAtom('atom1'))<1>
+  Adjoint(subbloq=TestAtom('atom2'))<0>.q -> q
+  q -> Adjoint(subbloq=TestAtom('atom0'))<2>.q
+--------------------
+Adjoint(subbloq=TestAtom('atom0'))<2>
+  Adjoint(subbloq=TestAtom('atom1'))<1>.q -> q
+  q -> RightDangle.reg"""
+    )
+
+
+def test_cbloq_adjoint_function():
+    cbloq = qlt_testing.assert_valid_bloq_decomposition(TestSerialCombo())
+    assert isinstance(cbloq, CompositeBloq)
+    adj_cbloq = _adjoint_cbloq(cbloq)
+    assert isinstance(adj_cbloq, CompositeBloq)
+    qlt_testing.assert_valid_cbloq(adj_cbloq)
+
+    assert (
+        adj_cbloq.debug_text()
+        == """\
+Adjoint(subbloq=TestAtom('atom2'))<0>
+  LeftDangle.reg -> q
+  q -> Adjoint(subbloq=TestAtom('atom1'))<1>.q
+--------------------
+Adjoint(subbloq=TestAtom('atom1'))<1>
+  Adjoint(subbloq=TestAtom('atom2'))<0>.q -> q
+  q -> Adjoint(subbloq=TestAtom('atom0'))<2>.q
+--------------------
+Adjoint(subbloq=TestAtom('atom0'))<2>
+  Adjoint(subbloq=TestAtom('atom1'))<1>.q -> q
+  q -> RightDangle.reg"""
+    )
+
+
+def test_adjoint_signature():
+    cnot = CNOT()
+    adj = Adjoint(cnot)  # specifically use the Adjoint wrapper for testing
+    assert cnot.signature == adj.signature  # all thru
+
+    zero = ZeroState()
+    adj = Adjoint(zero)  # specifically use the Adjoint wrapper for testing
+    assert len(zero.signature) == len(adj.signature)
+    (reg,) = zero.signature
+    (adj_reg,) = adj.signature
+    assert reg.name == adj_reg.name
+    assert reg.side == Side.RIGHT
+    assert adj_reg.side == Side.LEFT
+
+
+def test_adjoint_supports_decompose():
+    assert not CNOT().supports_decompose_bloq()
+    assert not Adjoint(CNOT()).supports_decompose_bloq()
+
+    assert CSwap(bitsize=5).supports_decompose_bloq()
+    assert Adjoint(CSwap(bitsize=5)).supports_decompose_bloq()
+
+
+def test_adjoint_adjoint():
+    zero = ZeroState()
+    adj = Adjoint(zero)  # specifically use the Adjoint wrapper for testing
+    assert adj.adjoint() == zero
+
+
+def test_build_call_graph():
+    n = sympy.Symbol('_n0')
+    bc = Adjoint(TestBloqWithCallGraph()).bloq_counts()
+    assert bc == {
+        Adjoint(subbloq=TestAtom()): n,
+        Adjoint(subbloq=TestSerialCombo()): 1,
+        Adjoint(subbloq=TestParallelCombo()): 1,
+    }
```

Comment:
Do we have a test bloq that requires more than one level of decomposition? E.g. I might be worried about the adjoint propagating all the way down the call graph.